Not sure if related but I got a panic on one of the managers (I didn't get that last time with a `-1000` score):

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x59e9ce]

goroutine 208644 [running]:
panic(0x1af02a0, 0xc82000a0d0)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/docker/docker/daemon.(*Daemon).SystemInfo(0xc8203d2680, 0x10, 0x0, 0x0)
	/go/src/github.com/docker/docker/daemon/info.go:108 +0xd5e
github.com/docker/docker/daemon/cluster/executor/container.(*executor).Describe(0xc820a80420, 0x7fc80e294010, 0xc82124d6e0, 0x0, 0x0, 0x0)
	/go/src/github.com/docker/docker/daemon/cluster/executor/container/executor.go:29 +0x6a
github.com/docker/swarmkit/agent.(*session).start(0xc822da9e00, 0x7fc80e294010, 0xc82124d6e0, 0x0, 0x0)
	/go/src/github.com/docker/docker/vendor/src/github.com/docker/swarmkit/agent/session.go:82 +0x1b6
github.com/docker/swarmkit/agent.(*session).run(0xc822da9e00, 0x7fc80e294010, 0xc82124d6e0, 0x21df1b7e)
	/go/src/github.com/docker/docker/vendor/src/github.com/docker/swarmkit/agent/session.go:58 +0x5e
created by github.com/docker/swarmkit/agent.newSession
	/go/src/github.com/docker/docker/vendor/src/github.com/docker/swarmkit/agent/session.go:51 +0x1f0
```

Seems more like an OOM than a nil pointer dereference looking at the specific line where it panics but just speculating.

Also all the Managers reverted back to worker nodes somehow after the panic. Will try with a few other runs and try to grab more infos if I can.

--- a/cmd/dockerd/daemon_unix.go
+++ b/cmd/dockerd/daemon_unix.go
@@ -61,6 +61,7 @@ func (cli *DaemonCli) setupConfigReloadTrap() {
 func (cli *DaemonCli) getPlatformRemoteOptions() []libcontainerd.RemoteOption {
 	opts := []libcontainerd.RemoteOption{
 		libcontainerd.WithDebugLog(cli.Config.Debug),
+		libcontainerd.WithOOMScore(cli.Config.OOMScoreAdjust),
 	}
 	if cli.Config.ContainerdAddr != "" {
 		opts = append(opts, libcontainerd.WithRemoteAddr(cli.Config.ContainerdAddr))

--- a/daemon/config_unix.go
+++ b/daemon/config_unix.go
@@ -34,6 +34,7 @@ type Config struct {
 	Ulimits              map[string]*units.Ulimit `json:"default-ulimits,omitempty"`
 	Runtimes             map[string]types.Runtime `json:"runtimes,omitempty"`
 	DefaultRuntime       string                   `json:"default-runtime,omitempty"`
+	OOMScoreAdjust       int                      `json:"oom-score-adjust,omitempty"`
 }
 
 // bridgeConfig stores all the bridge driver specific
@@ -90,6 +91,7 @@ func (config *Config) InstallFlags(cmd *flag.FlagSet, usageFn func(string) strin
 	config.Runtimes = make(map[string]types.Runtime)
 	cmd.Var(runconfigopts.NewNamedRuntimeOpt("runtimes", &config.Runtimes, stockRuntimeName), []string{"-add-runtime"}, usageFn("Register an additional OCI compatible runtime"))
 	cmd.StringVar(&config.DefaultRuntime, []string{"-default-runtime"}, stockRuntimeName, usageFn("Default OCI runtime to be used"))
+	cmd.IntVar(&config.OOMScoreAdjust, []string{"-oom-score-adjust"}, -500, usageFn("Set the oom_score_adj for the daemon"))
 
 	config.attachExperimentalFlags(cmd, usageFn)
 }

--- a/daemon/daemon.go
+++ b/daemon/daemon.go
@@ -434,7 +434,11 @@ func NewDaemon(config *Config, registryService registry.Service, containerdRemot
 		}
 	}
 
-	if err = setupDaemonRoot(config, realRoot, rootUID, rootGID); err != nil {
+	if err := setupDaemonRoot(config, realRoot, rootUID, rootGID); err != nil {
+		return nil, err
+	}
+
+	if err := setupDaemonProcess(config); err != nil {
 		return nil, err
 	}
 

--- a/daemon/daemon_solaris.go
+++ b/daemon/daemon_solaris.go
@@ -161,3 +161,7 @@ func (daemon *Daemon) setDefaultIsolation() error {
 func rootFSToAPIType(rootfs *image.RootFS) types.RootFS {
 	return types.RootFS{}
 }
+
+func setupDaemonProcess(config *Config) error {
+	return nil
+}

--- a/daemon/daemon_unix.go
+++ b/daemon/daemon_unix.go
@@ -1139,3 +1139,19 @@ func rootFSToAPIType(rootfs *image.RootFS) types.RootFS {
 		Layers: layers,
 	}
 }
+
+// setupDaemonProcess sets various settings for the daemon's process
+func setupDaemonProcess(config *Config) error {
+	// setup the daemons oom_score_adj
+	return setupOOMScoreAdj(config.OOMScoreAdjust)
+}
+
+func setupOOMScoreAdj(score int) error {
+	f, err := os.OpenFile("/proc/self/oom_score_adj", os.O_WRONLY, 0)
+	if err != nil {
+		return err
+	}
+	_, err = f.WriteString(strconv.Itoa(score))
+	f.Close()
+	return err
+}

--- a/daemon/daemon_windows.go
+++ b/daemon/daemon_windows.go
@@ -431,3 +431,7 @@ func rootFSToAPIType(rootfs *image.RootFS) types.RootFS {
 		BaseLayer: rootfs.BaseLayer,
 	}
 }
+
+func setupDaemonProcess(config *Config) error {
+	return nil
+}

--- a/docs/reference/commandline/dockerd.md
+++ b/docs/reference/commandline/dockerd.md
@@ -54,9 +54,10 @@ weight = -1
       --label=[]                             Set key=value labels to the daemon
       --log-driver="json-file"               Default driver for container logs
       --log-opt=[]                           Log driver specific options
-      --mtu=0                                Set the containers network MTU
       --max-concurrent-downloads=3           Set the max concurrent downloads for each pull
       --max-concurrent-uploads=5             Set the max concurrent uploads for each push
+      --mtu=0                                Set the containers network MTU
+      --oom-score-adjust=-500                Set the oom_score_adj for the daemon
       --disable-legacy-registry              Do not contact legacy registries
       -p, --pidfile="/var/run/docker.pid"    Path to use for daemon PID file
       --raw-logs                             Full timestamps without ANSI coloring
@@ -1057,6 +1058,7 @@ This is a full example of the allowed configuration options on Linux:
 	"insecure-registries": [],
 	"disable-legacy-registry": false,
 	"default-runtime": "runc",
+	"oom-score-adjust": -500,
 	"runtimes": {
 		"runc": {
 			"path": "runc"

--- a/libcontainerd/remote_linux.go
+++ b/libcontainerd/remote_linux.go
@@ -54,6 +54,7 @@ type remote struct {
 	runtimeArgs   []string
 	daemonWaitCh  chan struct{}
 	liveRestore   bool
+	oomScore      int
 }
 
 // New creates a fresh instance of libcontainerd remote.
@@ -402,7 +403,10 @@ func (r *remote) runContainerdDaemon() error {
 		return err
 	}
 	logrus.Infof("New containerd process, pid: %d", cmd.Process.Pid)
-
+	if err := setOOMScore(cmd.Process.Pid, r.oomScore); err != nil {
+		utils.KillProcess(cmd.Process.Pid)
+		return err
+	}
 	if _, err := f.WriteString(fmt.Sprintf("%d", cmd.Process.Pid)); err != nil {
 		utils.KillProcess(cmd.Process.Pid)
 		return err
@@ -415,6 +419,16 @@ func (r *remote) runContainerdDaemon() error {
 	}() // Reap our child when needed
 	r.daemonPid = cmd.Process.Pid
 	return nil
+}
+
+func setOOMScore(pid, score int) error {
+	f, err := os.OpenFile(fmt.Sprintf("/proc/%d/oom_score_adj", pid), os.O_WRONLY, 0)
+	if err != nil {
+		return err
+	}
+	_, err = f.WriteString(strconv.Itoa(score))
+	f.Close()
+	return err
 }
 
 // WithRemoteAddr sets the external containerd socket to connect to.
@@ -509,4 +523,19 @@ func (l liveRestore) Apply(r Remote) error {
 		return nil
 	}
 	return fmt.Errorf("WithLiveRestore option not supported for this remote")
+}
+
+// WithOOMScore defines the oom_score_adj to set for the containerd process.
+func WithOOMScore(score int) RemoteOption {
+	return oomScore(score)
+}
+
+type oomScore int
+
+func (o oomScore) Apply(r Remote) error {
+	if remote, ok := r.(*remote); ok {
+		remote.oomScore = int(o)
+		return nil
+	}
+	return fmt.Errorf("WithOOMScore option not supported for this remote")
 }

